### PR TITLE
v1.5.0

### PR DESCRIPTION
**Pull request by frankkopp**
# PR: Merge `dev_1.5` → `master` — FrankyCPP v1.5.0

## Summary

FrankyCPP v1.5 delivers **+103.7 ELO** over v1.4 and now plays **on par with Stockfish 18 at 2700 ELO** (+6.9 ELO). Key improvements: transposition table 4-way bucket redesign, Lazy SMP best-thread selection, comprehensive namespace cleanup, Search refactoring, and a major Arena testing infrastructure overhaul. 47 commits, 319 files changed (184 C++ source files, ~20K lines inserted, ~16.5K deleted).

---

## Match Results (4 threads, 300s/game, 100 games each)

| Opponent                | Score     | W/D/L              | ELO Δ    |
|-------------------------|-----------|--------------------|----------|
| **Stockfish 18 (2700)** | **51–49** | **49W / 4D / 47L** | **+6.9** |
| FrankyCPP v1.4          | 64.5–35.5 | 46W / 37D / 17L    | +103.7   |
| FrankyCPP v1.3          | 75.5–24.5 | 62W / 27D / 11L    | +195.5   |

Test suites: **1747/2875** (60.7%)

---

### ♟️ Transposition Table: 4-Way Bucket Refactor
- Refactored TT from single-entry per slot to **4-way associative buckets**, improving hit rates and reducing collision evictions
- Hardened SMP probe safety for concurrent access
- Enhanced torn-read detection and reporting in concurrent tests
- Updated default TT size to **256 MB**, new default pawn TT size

### 🧵 Best Thread Selection for Lazy SMP (+11.2 ELO)
- **Best-thread selection**: after search, picks the best result across all SMP threads instead of always using thread 0
- Depth+score heuristic with configurable `BEST_THREAD_SCORE_MARGIN` (default 50cp) — deeper threads win unless a shallower thread has a significantly better score
- Final UCI info line correction when a non-main thread is selected
- Guards for hard-limited searches, TB root probing, and mate detection edge cases
- New UCI options: `USE_BEST_THREAD_SELECTION`, `BEST_THREAD_SCORE_MARGIN`
- Validated at **+11.2 ELO** (103.7 vs v1.4, up from 92.5 baseline; +6.9 vs Stockfish 2700, up from −17.4 baseline)

### 🏗️ Namespace Refactor (Complete)
- Removed transitional `using namespace chess;` from `types.h`
- Added scoped `using namespace chess;` inside namespace blocks across all headers (engine, openingbook, enginetest, engine_arena, tablebase)
- Created `chesscore/fwd.h` with centralized forward declarations
- Replaced scattered forward declarations with the centralized header

### 🔧 Search Code Refactoring (6 commits)
Systematic cleanup of `Search::run()` and related code for maintainability:
- Extract `resetSearchState()` — per-search init of Search-owned state
- Extract `TBRootInfo` struct — groups 4 loose TB root fields with `reset()`
- Consolidate `SearchThreadData::resetForNewSearch()` / `reset()` — replace 33-line and 11-line inline loops
- Extract `joinHelperThreads()` — consolidate 3 inline join-and-clear loops
- Extract `joinTimerThread()` — mutex-guarded timer cleanup
- Extract `applyTBRootOverride()` and `extractPonderMove()` — ~100 lines moved out of `run()`

### 🏟️ Arena Testing Infrastructure
- **Unified configuration**: replace `testSuites` with `testSuiteRuns` format, per-suite overrides (timePerMove, maxDepth)
- **Tag system**: tag field on all config types, propagated through runners to result JSON
- **Engine summary** (`--summary`): aggregated test suite/match/benchmark results per engine with `--history` flag
- **`--show-config`**: query engine UCI options from Arena CLI
- **Refactor stages 12-13**: code review findings and fixes (fixedWidth edge case, namespace cleanup, `[[nodiscard]]`, utility consolidation)
- **Bug fixes**: duplicate/phantom engine entries in baseline reports, engine name inconsistency across result sources
- 42 new test result JSON files (baseline runs for v1.1–v1.5, FrankyGo, Stockfish)

### 📦 Build System
- **CMake GLOB refactor**: replace manually maintained file lists with `file(GLOB CONFIGURE_DEPENDS)` — new source files auto-discovered
- Rename `SRCS_CONFIG`/`TEST_CONFIG` to avoid CMake naming collision
- Move `SearchTreeSizeTest` production exclusion from CMake to `#ifndef` source guards
- Add Chess-EPDs submodule

### 📄 Documentation & Plans
- Architecture.md: Lazy SMP overview diagram, threading detail, PawnTT
- Build Guide, CLion Setup, Sanitizer Guide: updated binary names for v1.5
- New plans: Move Ordering & Pruning Improvements, MultiPV (v1.6+), TT Performance Optimization
- Archived completed plans
- Updated CLAUDE.md and copilot-instructions.md

---

## Breaking Changes

None. UCI protocol fully backward compatible. New UCI options have sensible defaults.

---

## Testing

- All existing unit tests pass
- New tests: `BestThreadSelectionEnabled`, `BestThreadSelectionDisabled`, `selectBestThread` (7 synthetic scenarios)
- Enhanced TT concurrent test with torn-read detection
- SearchTreeSizeTest updated with threading support and SEE quiet pruning
- Arena validation: 100-game matches at 300s/game against Stockfish 2700 and previous versions

---

## New Configuration Options

| Option                      | Type | Default | Description                                     |
|-----------------------------|------|---------|-------------------------------------------------|
| `USE_BEST_THREAD_SELECTION` | bool | `true`  | Enable best-thread selection across SMP threads |
| `BEST_THREAD_SCORE_MARGIN`  | int  | `50`    | Score margin (cp) for preferring deeper thread  |
| `TT_SIZE_MB`                | int  | `256`   | Default TT size (was 128)                       |

### CODE REVIEW

File: .github/copilot-instructions.md [hunk 1]
```diff
@@ -452,6 +452,8 @@ $env:VAR = "value"
 
 **CRITICAL: Always ask user before committing! Never commit without explicit permission.**
 
+**CRITICAL: Do NOT use the terminal to determine status or diff or anything - YOU HAVE A BIG BUG and can't read the terminal's git output anyway. Just do the commit!!**
```
[the remark this answers is not in the document]
Comment (frankkopp) [reply .github/copilot-instructions.md]:
```suggestion
```
